static/js/components/ObservationPlanRequestForm.jsx

```diff
@@ -0,0 +1,288 @@
+  },
+  allocationSelect: {
+    width: "100%",
+  },
+  localizationSelect: {
+    width: "100%",
+  },
+  allocationSelectItem: {
+    whiteSpace: "break-spaces",
+  },
+  localizationSelectItem: {
+    whiteSpace: "break-spaces",
+  },
+  container: {
+    width: "99%",
+    marginBottom: "1rem",
+  },
+}));
+
+const ObservationPlanRequestForm = ({ gcnevent }) => {
+  const classes = useStyles();
+  const dispatch = useDispatch();
+
+  const { telescopeList } = useSelector((state) => state.telescopes);
+  const { allocationList } = useSelector((state) => state.allocations);
+
+  const allGroups = useSelector((state) => state.groups.all);
+  const [selectedAllocationId, setSelectedAllocationId] = useState(null);
+  const [selectedGroupIds, setSelectedGroupIds] = useState([]);
+  const [selectedLocalizationId, setSelectedLocalizationId] = useState(null);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+
+  const { instrumentList, instrumentFormParams } = useSelector(
+    (state) => state.instruments
+  );
+
+  useEffect(() => {
+    const getAllocations = async () => {
+      // Wait for the allocations to update before setting
+      // the new default form fields, so that the allocations list can
+      // update
+      const result = await dispatch(
+        allocationActions.fetchAllocations({
+          apitype: "api_observationplan_classname",
+        })
+      );
+
+      const { data } = result;
+      setSelectedAllocationId(data[0]?.id);
+      setSelectedGroupIds([data[0]?.group_id]);
+      setSelectedLocalizationId(gcnevent.localizations[0]?.id);
+    };
+
+    getAllocations();
+
+    dispatch(
+      instrumentActions.fetchInstrumentForms({
+        apitype: "api_observationplan_classname",
+      })
+    );
+
+    // Don't want to reset everytime the component rerenders and
+    // the defaultStartDate is updated, so ignore ESLint here
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [
+    dispatch,
+    setSelectedAllocationId,
+    setSelectedGroupIds,
+    setSelectedLocalizationId,
+  ]);
+
+  // need to check both of these conditions as selectedAllocationId is
+  // initialized to be null and useEffect is not called on the first
+  // render to update it, so it can be null even if allocationList is not
+  // empty.
+  if (
+    allocationList.length === 0 ||
+    !selectedAllocationId ||
+    Object.keys(instrumentFormParams).length === 0
+  ) {
+    return <h3>No robotic instruments available...</h3>;
+  }
+
+  if (
+    !allGroups ||
+    allGroups.length === 0 ||
+    telescopeList.length === 0 ||
+    instrumentList.length === 0
+  ) {
+    return (
+      <div>
+        <CircularProgress color="secondary" />
+      </div>
+    );
+  }
+
+  if (
+    !allGroups ||
+    allGroups.length === 0 ||
+    telescopeList.length === 0 ||
+    instrumentList.length === 0
+  ) {
+    return (
+      <div>
+        <CircularProgress color="secondary" />
+      </div>
+    );
+  }
+
```

Comment:
duplicate code.

Reply:
This doesn't just make it doubly good? ;)